### ISSUE TITLE
SD Card is not reconized on power up 1.1.9

### ISSUE THREAD

**Issue by Rgatliff** [MarlinFirmware/Marlin#11533]
I am running a Anet A8 3D Printer.  I just upgraded from Marlin 1.1.8 to the new Marlin 1.1.9 release.  I am using the standard configuration files from the **Example Confirguartion** folder for Anet A8

Under Marlin 1.1.9, each time that I power on my printer, the menu shows “No SD Card”.  I have to select the “Init. SD Card” command to get it to display “Print from SD Card”.  As long as the printer stays powered on, the SD card functions as they under Marlin 1.1.8 and I do not have to select the “Init. SD card” command again. 

Under Marlin 1.1.8 I did not have this problem, it recognized my SD card on power up.
[Configuration.zip](https://github.com/MarlinFirmware/Marlin/files/2281369/Configuration.zip)

**Comment by roel8032** [comment]
I have the same issue with the Anet A6.

Rgatiff please try to uncomment in your config file:
//#define SD_CHECK_AND_RETRY

By the way, not working for me.

**Comment by cjsoong** [comment]
@Rgatliff @roel8032  
This problem will only appear on a few mainboard. Because there is no SD detect pin.
You can try this change to temporarily solve your problem.
#### Marlin_main.cpp
```cpp
/* power on init. sd card */
#if ENABLED(SDSUPPORT)
  if (!card.cardOK) card.initsd();
#endif
```
![sd](https://user-images.githubusercontent.com/26795599/44011000-75426b7a-9ee8-11e8-88a2-5677da5fadd3.jpg)

**Comment by roel8032** [comment]
Not working:
Error:
exit status 1
'if' was not declared in this scope.

**Comment by cjsoong** [comment]
@roel8032 
I compiled is Ok.
You can change it to try this
```cpp
#if ENABLED(SDSUPPORT)
  card.initsd();
#endif
```

**Comment by thinkyhead** [comment]
Have you tried various SD cards? Have you tried doing a "low level format" of the SD cards to see if it improves their response time?

**Comment by roel8032** [comment]
Thanks cjsoong. This works.

**Comment by roel8032** [comment]
@Thinkyhead, how to low lever format an SD card?
I used SD formatter.
Besides the card is known by a Infitary M508 3d printer, and also with Anet A6 at 1.1.8.

**Comment by thinkyhead** [comment]
I've never done a low-level format. I'd have to Google it.

Seems the patch works, though, which could indicate an actual issue. The way we can find out is to add some debugging…

```cpp
/* power on init. sd card */
#if ENABLED(SDSUPPORT)
  SERIAL_ECHOLNPAIR("At startup the card is ", card.cardOK ? "OK" : "not OK");
  card.initsd();
#endif
```

**Comment by Rgatliff** [comment]
Thank you, I will update using this code.

**Comment by Rgatliff** [comment]
I installed the code at the beginning of the main program loop in Marlin_main.cpp and now on power up, the SD drive is recognized.  I appreciate you help on this.

**Comment by Rgatliff** [comment]
As it turns out, your code fixed the CD Init issue but it created another issue.  Now when the printer is waiting on work to do,  the exhaust fan is creating a vibration due to a very low current is going to it but not enough to cause the fan to turn over causing the fan to set there and vibration.  After I backed the code out, the exhaust fan quite make a noise.  Should I have put your code somewhere in the `Marlin_main.cpp`?

**Comment by roel8032** [comment]
That is quite normal for a Chinese fan. It is the bearing that makes noise. Touch it with your finger

**Comment by cjsoong** [comment]
@Rgatliff
 I have checked the code of "initsd" and it doesn't matter with FAN.
I have no problem with my machine.
Are you sure that it is related to adding "initsd"?

**Comment by roel8032** [comment]
Not related.

**Comment by Rgatliff** [comment]
I have a feeling that that somehow it is related to where I put it and not due to the code that I inserted which is causing some other code to get disrupted.

  

I inserted the code right after the following 

/* The main Marlin program loop

…

….

void loop() {

#if ENABLED(SDSUPPORT)          

     card.initsd();

#endif

 

Should I have inserted it right before this main program loop code?

**Comment by tommyvhb** [comment]
@Rgatliff 
> Should I have inserted it right before this main program loop code? 

Yes, the provided code should be inserted before the main loop. This way the SD card is only initialised at power up.
In your case the card is constantly being initialised. When the card is initialised, the cooling fan gets a spin and that's what you're seeing because the fan is activated for a very short time at each initialisation.

@cjsoong 
No problems here with this sollution. Thank you.

**Comment by Rgatliff** [comment]
That is what it needed.  Put it at the end of the setup routine. Now the fan is not making a noise.

Thanks to all the help me with this issue.

Robert

**Comment by del-vincent** [comment]
The solution is here below:

There is an error in cardreader.h on line 207 
replace:
#if PIN_EXISTS(SD_DETECT)
by:
#if PIN_EXISTS(SD_DETECT_PIN)

**Comment by AnHardt** [comment]
No.
```
// Macros to support pins/buttons exist testing
#define _PINEX_1(PN)        (defined(PN##_PIN) && PN##_PIN >= 0)
#define PIN_EXISTS(V...)    DO(PINEX,&&,V)
```
_PIN is added automagically.
